cpp/src/parquet/column_chunker.cc

```diff
@@ -0,0 +1,778 @@
+  uint64_t avg_size = (min_size + max_size) / 2;
+  // we skip calculating gearhash for the first `min_size` bytes, so we are looking for
+  // a smaller chunk as the average size
+  uint64_t target_size = avg_size - min_size;
+  size_t mask_bits = static_cast<size_t>(std::floor(std::log2(target_size)));
+  // -3 because we are using 8 hash tables to have more gaussian-like distribution
+  // `norm_factor` narrows the chunk size distribution aroun avg_size
+  size_t effective_bits = mask_bits - 3 - norm_factor;
+  return std::numeric_limits<uint64_t>::max() << (64 - effective_bits);
+}
+
+ContentDefinedChunker::ContentDefinedChunker(const LevelInfo& level_info,
+                                             std::pair<uint64_t, uint64_t> size_range,
+                                             uint8_t norm_factor)
+    : level_info_(level_info),
+      min_size_(size_range.first),
+      max_size_(size_range.second),
+      hash_mask_(GetMask(size_range.first, size_range.second, norm_factor)) {}
+
+template <typename T>
+void ContentDefinedChunker::Roll(const T value) {
+  constexpr size_t BYTE_WIDTH = sizeof(T);
+  chunk_size_ += BYTE_WIDTH;
+  if (chunk_size_ < min_size_) {
+    // short-circuit if we haven't reached the minimum chunk size, this speeds up the
+    // chunking process since the gearhash doesn't need to be updated
+    return;
+  }
+  auto bytes = reinterpret_cast<const uint8_t*>(&value);
+  for (size_t i = 0; i < BYTE_WIDTH; ++i) {
+    rolling_hash_ = (rolling_hash_ << 1) + GEAR_HASH_TABLE[nth_run_][bytes[i]];
+    if ((rolling_hash_ & hash_mask_) == 0) {
+      has_matched_ = true;
+    }
+  }
+}
+
+void ContentDefinedChunker::Roll(std::string_view value) {
+  chunk_size_ += value.size();
+  if (chunk_size_ < min_size_) {
+    // short-circuit if we haven't reached the minimum chunk size, this speeds up the
+    // chunking process since the gearhash doesn't need to be updated
+    return;
+  }
+  for (char c : value) {
+    rolling_hash_ =
+        (rolling_hash_ << 1) + GEAR_HASH_TABLE[nth_run_][static_cast<uint8_t>(c)];
+    if ((rolling_hash_ & hash_mask_) == 0) {
+      has_matched_ = true;
+    }
+  }
+}
+
+bool ContentDefinedChunker::Check() {
+  // decide whether to create a new chunk based on the rolling hash; has_matched_ is
+  // set to true if we encountered a match since the last Check() call
+  if (ARROW_PREDICT_FALSE(has_matched_)) {
+    has_matched_ = false;
+    // in order to have a normal distribution of chunk sizes, we only create a new chunk
+    // if the adjused mask matches the rolling hash 8 times in a row, each run uses a
+    // different gearhash table (gearhash's chunk size has exponential distribution, and
+    // we use central limit theorem to approximate normal distribution)
+    if (ARROW_PREDICT_FALSE(++nth_run_ >= 7)) {
+      nth_run_ = 0;
+      chunk_size_ = 0;
+      return true;
+    }
+  }
+  if (ARROW_PREDICT_FALSE(chunk_size_ >= max_size_)) {
```

Comment:
It seems that some chunks returned by `Calculate()` may be bigger than `max_size_`. We may need to check this in `Roll()` or before calling `Role()`.

Reply:
You are right, max chunk size is checked after updating the rolling hash, it works similarly to `max_page_size` which is also being checked after appending the value. 

If we were checking the max chunk size before updating the rolling hash, in case of a large record we may encounter the case that single record doesn't fit in max_chunk_size so it would endlessly try to close the chunk and fit the record to a new one where it doesn't fit again. 

This could be properly implemented of course but would require more complicated branches, also enforcing the record boundary would be harder. I think this is a trade-off we should keep, at least for now. Maybe in a follow-up I can properly enforce the max chunk size with more tests and some benchmarks.